pkg/proc/native/proc_darwin.c

```diff
@@ -133,67 +133,133 @@ thread_count(task_t task) {
 	return count;
 }
 
+
+const char* exception_string(int code) {
+	if (code == EXC_SOFTWARE) {
+		return "EXC_SOFTWARE";
+	} else if (code == EXC_BREAKPOINT) {
+		return "EXC_BREAKPOINT";
+	} else if (code == EXC_BAD_ACCESS) {
+		return "EXC_BAD_ACCESS";
+	}
+
+	return NULL;
+}
+
+const char* signal_string(int signal) {
+	if (signal == SIGHUP) {
+		return "SIGHUP";
+	} else if (signal == SIGKILL) {
+		return "SIGKILL";
+	} else if (signal == SIGSEGV) {
+		return "SIGSEGV";
+	} else if (signal == SIGSTOP) {
+		return "SIGSTOP";
+	} else if (signal == SIGURG) {
+		return "SIGURG";
+	}
+
+	return NULL;
+}
+
 mach_port_t
 mach_port_wait(mach_port_t port_set, task_t *task, int nonblocking) {
 	kern_return_t kret;
-	thread_act_t thread;
-	NDR_record_t *ndr;
-	integer_t *data;
-	union
-	{
-		mach_msg_header_t hdr;
-		char data[256];
-	} msg;
+	int64_t *data;
 	mach_msg_option_t opts = MACH_RCV_MSG|MACH_RCV_INTERRUPT;
 	if (nonblocking) {
 		opts |= MACH_RCV_TIMEOUT;
 	}
 
+	char req[128];
+
 	// Wait for mach msg.
-	kret = mach_msg(&msg.hdr, opts,
-			0, sizeof(msg.data), port_set, 10, MACH_PORT_NULL);
+	kret = mach_msg((mach_msg_header_t*) &req, opts,
+			0, sizeof(req), port_set, 10, MACH_PORT_NULL);
 	if (kret == MACH_RCV_INTERRUPTED) return kret;
 	if (kret != MACH_MSG_SUCCESS) return 0;
 
+	__Request__mach_exception_raise_t* msg = (__Request__mach_exception_raise_t*) req;
+	// TODO: we actually want to propagate the task back to Go, but this seems to randomly hang
+	// *task = msg->task.name;
```

Comment:
Are there cases where we wouldn't want to?

Reply:
Well, propagating the task back to Go has one main use case: Getting an information from the inferior, if it uses `execve`. Previously, this was needed because of the exec/fork-flow, which was quite unstable and susceptible to race conditions. Now, the only use case that remains would be, if an inferior itself uses `execve` so the underlying task would change and we want to propagate this back to Go.

I can have another look at this, the whole function is not ideal, since it is recursively done in C. While I doubt that there is a large number of recursions, this is still not ideal.

A "cleaner" way would probably be to return to Go for each exception, have a look at the exception in Go and decide to wait again there. This could however be a little bit slower because of additional context switches.

Reply:
Ack, makes sense.

I agree with tossing things over the fence back to Go for inspection, etc. I'm probably fine with taking the performance penalty there for the sake of added safety and correctness.

Reply:
I addressed this with a goto/label approach and also the task propagation seems to work now after my recent commit